enable 3D picking with perspective camera

beanli161514 opened the pull request:
# References and relevant issues

Closes #([issue-6751](https://github.com/napari/napari/issues/6751))

# Description

Change some part of the calculation of event.position and event.view_direction. Now get_value() and get_ray_intersection() works normally with perspective camera. So functions like hovering to show properties in point layer works fine now.

![demo](https://github.com/user-attachments/assets/58036c3c-5776-4f13-bb49-562334b53834)

Commit:
enable picking with perspective camera

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -634,6 +639,33 @@ def _add_overlay_to_visual(self, overlay: Overlay) -> None:
+        p1 = [
+            w / 2,
+            h / 2,
+            -1e10,
```
Any reason why we shouldn't use actual infinite? Or anything bigger than `1e10`?

beanli161514 replied:
I've tried float(np.inf), transform.imap can't handle that. Actually anything bigger than 10e6 would yeild enough precision.

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -332,8 +332,12 @@ def _map_canvas2world(
         nd = self.viewer.dims.ndisplay
-        transform = self.view.scene.transform
-        mapped_position = transform.imap(list(position))[:nd]
+        transform = self.view.camera._scene_transform
```
what's the difference between `view.scene.transform` and `camera._scene_transform`? Shouldn't they be the same?

beanli161514 replied:
I don't exactly know, here's my guess. view.scene.transform maps the coordinates from visual(data) to scene, camera._scene_transform transforms scene coordinates to camera coordinates and then project them to canvas.

brisvag replied:
I don't think that's right, I just tried to revert to `self.view.scene.transform` and everything seems to be working fine, also in perspective mode.

Was there a specific reason why you made this change? Unless there's a reason I'm missing, I think it's better to go back and not use private attributes:

```suggestion
        transform = self.view.scene.transform
```

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -332,8 +332,12 @@ def _map_canvas2world(
+        mapped_position = transform.imap(list(position))
+        mapped_position = (
+            mapped_position[0:nd] / mapped_position[nd]
+        )  # cartesian to homogeneous coordinates
```
I would put comments above the code for readability (should prevent this multiline nonsense from the black formatter):

```suggestion
        # cartesian to homogeneous coordinates
        mapped_position = mapped_position[0:nd] / mapped_position[nd]
```

Same thing happens lower down!

brisvag replied:
Now testing locally, this is wrong. It breaks in 2D, because it makes no sense in 2D (and ends up dividing by zero). This should only happen in 3D mode!

beanli161514 replied:
I think in 2D, the coordinates are not homogeneous, thus the last column is left 0.

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -382,9 +386,10 @@ def _process_mouse_event(
-        )
+        # event.view_direction = self.viewer.camera.calculate_nd_view_direction(
+        #     self.viewer.dims.ndim, self.viewer.dims.displayed
+        # )
```
```suggestion
```

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -382,9 +386,10 @@ def _process_mouse_event(
+        #     self.viewer.dims.ndim, self.viewer.dims.displayed
+        # )
+        event.view_direction = self._calculate_view_direction(event.pos)
         event.up_direction = self.viewer.camera.calculate_nd_up_direction(
             self.viewer.dims.ndim, self.viewer.dims.displayed
         )
```
Is it possible that this method also needs to be updated similarly?

beanli161514 replied:
I am not familiar with this fucntion, it seems have something to do with the switch between 2D and 3D mode.

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -634,6 +639,33 @@ def _add_overlay_to_visual(self, overlay: Overlay) -> None:
             vispy_overlay.node.parent = self.view.scene
         self._overlay_to_visual[overlay] = vispy_overlay
 
+    def _calculate_view_direction(self, event_pos) -> List[float]:
```
Nice work! Makes sense to me :) For readability, would be nice to have more explicit variable names instead of relying on comments, like:

`p0` -> `pos_scene`
`p1` -> `eye_scene`

and so on :)

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -634,6 +639,33 @@ def _add_overlay_to_visual(self, overlay: Overlay) -> None:
+        d = d[0:nd]
+        d = d / np.linalg.norm(d)
+
+        p0 = list(p0[::-1])  # xyz to zyx
```
unused, can remove

```suggestion
```

brisvag reviewed (commented):
Awesome stuff! Thanks for the PR, finally someone got around to tackling this :muscle: 

A few comments/questions in the review.

kevinyamauchi commented on napari/_vispy/canvas.py:
```diff
@@ -382,9 +386,10 @@ def _process_mouse_event(
+        # event.view_direction = self.viewer.camera.calculate_nd_view_direction(
+        #     self.viewer.dims.ndim, self.viewer.dims.displayed
+        # )
+        event.view_direction = self._calculate_view_direction(event.pos)
```
I think this will need to be converted into nD world coordinates. I think as-is, this will be in the 3D coordinates of the rendered scene. (not totally sure - didn't look too closely)

brisvag replied:
Yeah, I think this is the main remaining hurdle.

Basically, this code with a 4D image is currently breaking (testing by just moving mouse on the canvas with #7126 changes included):

```
import napari
from skimage.data import binary_blobs

blobs = binary_blobs(
    length=64,
    blob_size_fraction=0.2,
    n_dim=4,
    volume_fraction=0.1,
)
v = napari.Viewer(ndisplay=3)
v.add_image(blobs)
```

Essentially we need the 3D view direction to be embedded in an nD vector so it can be properly sliced later.

Commit:
enable dragging with perspective camera

Commit:
enable dragging with perspective camera

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -634,6 +640,55 @@ def _add_overlay_to_visual(self, overlay: Overlay) -> None:
+        d = list(d[::-1])
+        return d
+
+    def _adjust_event_position(
```
```suggestion
    def _move_event_to_camera_center_plane(
```

brisvag reviewed (commented):
Looking good! Still a couple things to fix, but we're close.


Then, we need to look through tests and see if anything else other than the ND view ray problem needs fixing. We should also consider adding some new test.

beanli161514 committed:
Update napari/_vispy/canvas.py

Yeah there's no difference between these two transformations. The problem was caused by the incorrect calculation of view direction, not the transformation.

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

beanli161514 committed:
Update napari/_vispy/canvas.py

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Commit:
Sovling nd cases

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

brisvag commented on napari/_vispy/canvas.py:
```diff
@@ -634,6 +640,64 @@ def _add_overlay_to_visual(self, overlay: Overlay) -> None:
+        eye_scene = eye_scene[0:nd] / eye_scene[nd]
+
+        # calculate direction of the ray
+        d = eye_scene - click_scene
```
This is the problem @jni:

```suggestion
        d = click_scene - eye_scene
```

beanli161514 committed:
Update napari/_vispy/canvas.py

Co-authored-by: Lorenzo Gaifas <brisvag@gmail.com>

Czaki committed:
Merge branch 'main' into main

Czaki committed:
Merge branch 'main' into main

brisvag commented on .gitignore:
```diff
@@ -175,3 +175,6 @@ napari/resources/icons/_themes/
 github_cache.sqlite
+
+.gitignore
+/tests
```
Was this on purpose? Seems wrong...

beanli161514 replied:
Oh that's a mistake, I forgot the changes I made to .gitignore to add another custom tests folder

brisvag reviewed (commented):
For testing: using `Layer._get_value_3d()` is actually useless, cause the inputs of that method are already "sanitized" from the perspective component.

So we need to somehow just have some test on the canvas, where we actually set the FOV, get put an event in, and get the "proper" direction based on the FOV... only sensible way seems to just compute some values with this PR and harcode them in the test, assuming that this is *Correct*.

beanli161514 committed:
remove changes made to .gitignore

beanli161514 committed:
Merge branch 'napari:main' into main

Commit:
discard _move_event_to_camear_center_plane

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

Commit:
remove deprecated type annotations

Commit:
remove deprecated type annotations

beanli161514 committed:
Merge branch 'main' into main

Commit:
remove unused function
